src/lib.rs

```diff
@@ -725,16 +733,70 @@ impl<M: ManageConnection> Pool<M> {
         if broken {
             let _ = drop_connections(&inner, locked, vec![conn]).await;
         } else {
-            let conn = IdleConn::make_idle(Conn {
-                conn: conn,
-                birth: birth,
-            });
+            let conn = IdleConn::make_idle(Conn { conn, birth });
             locked.put_idle_conn(conn);
         }
 
         r.map_err(|e| RunError::User(e))
     }
 
+    /// Return connection back in to the pool
+    fn put_back(&self, birth: Instant, mut conn: Conn<M::Connection>) {
+        let inner = self.inner.clone();
+
+        // Supposed to be fast, but do it before locking anyways.
+        let broken = inner.manager.has_broken(&mut conn.conn);
+
+        futures::executor::block_on(async {
+            let mut locked = inner.internals.lock().await;
+            if broken {
+                let _ = drop_connections(&inner, locked, vec![conn.conn]).await;
+            } else {
+                let conn = IdleConn::make_idle(Conn {
+                    conn: conn.conn,
+                    birth,
+                });
+                locked.put_idle_conn(conn);
+            }
+        })
+    }
+
+    /// Retrieves a connection from the pool.
+    pub async fn get(&self) -> Result<PooledConnection<M>, RunError<M::Error>> {
+        let inner = self.inner.clone();
+
+        match get_idle_connection(inner).await {
```

Comment:
Same thing here: this code should be factored out so that `run()` and `get()` can share the same code path.